aws/resource_aws_instance.go

```diff
@@ -1260,6 +1254,95 @@ func resourceAwsInstanceUpdate(d *schema.ResourceData, meta interface{}) error {
 		}
 	}
 
+	if d.HasChange("root_block_device.0") && !d.IsNewResource() {
+		instance, err := resourceAwsInstanceFindByID(conn, d.Id())
+		if err != nil {
+			return fmt.Errorf("error retrieving instance %q: %w", d.Id(), err)
+		}
+		blockDevices, err := readBlockDevicesFromInstance(instance, conn)
+		if err != nil {
+			return fmt.Errorf("error retrieving volumes for instance %q: %w", d.Id(), err)
+		}
+		rd, ok := blockDevices["root"]
+		if !ok {
+			return fmt.Errorf("root volume no found for instance %q", d.Id())
+		}
+
+		rootDevice := rd.(map[string]interface{})
+		volumeID := rootDevice["volume_id"].(string)
+		input := ec2.ModifyVolumeInput{
+			VolumeId: aws.String(volumeID),
+		}
+		modifyVolume := false
+
+		rbd := d.Get("root_block_device.0").(map[string]interface{})
+		if v, ok := rbd["volume_size"]; ok {
+			currSize := rootDevice["volume_size"].(int64)
+			volumeSize := int64(v.(int))
+			if volumeSize != currSize {
```

Comment:
The existing root block device information can be fetched via `d.GetChange()` or to skip all this for each attribute, you can use `d.HasChange()` with the nested attributes, e.g. `d.HasChange("root_block_device.0.volume_size")`